### PR TITLE
feat(bulk_load): add bulk load manager for meta

### PR DESCRIPTION
**Pull request by GiantKing**

### CODE REVIEW

File: src/meta/meta_bulk_load_service.cpp [hunk 1]
```diff
@@ -101,6 +111,8 @@ void bulk_load_service::on_start_bulk_load(start_bulk_load_rpc rpc)
              request.file_provider_type,
              request.remote_root_path);
 
+    // clean old bulk load result
```
Comment (levy5307) [inline src/meta/meta_bulk_load_service.cpp:114]: clean --> clear or clean up

File: src/meta/meta_bulk_load_service.cpp [hunk 2]
```diff
@@ -234,21 +246,31 @@ void bulk_load_service::create_app_bulk_load_dir(const std::string &app_name,
     ainfo.cluster_name = req.cluster_name;
     ainfo.file_provider_type = req.file_provider_type;
     ainfo.remote_root_path = req.remote_root_path;
-    blob value = dsn::json::json_forwarder<app_bulk_load_info>::encode(ainfo);
+    ainfo.is_ever_ingesting = false;
+    ainfo.bulk_load_err = ERR_OK;
```
Comment (levy5307) [inline src/meta/meta_bulk_load_service.cpp:251]:
```
auto bulk_load_path = get_app_bulk_load_path(app_id);
  _meta_svc->get_meta_storage()->delete_node_recursively(
        bulk_load_path, [this, rpc, ainfo]() {
            ddebug_f("remove app({}) bulk load dir {} succeed", ainfo.app_name, bulk_load_path);
......
```

Comment (GiantKing) [reply src/meta/meta_bulk_load_service.cpp:251]: "delete_node_recursively" need a r-value, so I can't pass bulk_load_path in to lambda.

File: src/meta/meta_bulk_load_service.h [hunk 3]
```diff
@@ -40,13 +40,17 @@ struct app_bulk_load_info
     std::string file_provider_type;
     bulk_load_status::type status;
     std::string remote_root_path;
+    bool is_ever_ingesting;
+    error_code bulk_load_err;
```
Comment (hycdong) [inline src/meta/meta_bulk_load_service.h:44]: I suggest using `bulk_load_result`

File: src/meta/meta_bulk_load_service.h [hunk 4]
```diff
@@ -392,6 +404,16 @@ class bulk_load_service
         }
     }
 
+    inline error_code get_app_bulk_load_err_unlocked(int32_t app_id) const
```
Comment (hycdong) [inline src/meta/meta_bulk_load_service.h:407]: Same suggestion, using `get_app_bulk_load_result_unlocked`

File: src/meta/meta_bulk_load_service.h (hunk 3 above)
Comment (GiantKing) [reply src/meta/meta_bulk_load_service.h:44]: This param is useful only if status=bulk_load_status::BLS_FAILED, so it stand for "error", not "result"

File: src/meta/meta_bulk_load_service.h (hunk 4 above)
Comment (GiantKing) [reply src/meta/meta_bulk_load_service.h:407]: Same as above

File: src/common/bulk_load.thrift [hunk 5]
```diff
@@ -206,4 +206,5 @@ struct query_bulk_load_response
     // detailed bulk load state for each replica
     6:list<map<dsn.rpc_address, partition_bulk_load_state>> bulk_load_states;
     7:optional string                                       hint_msg;
+    8:bool                                                  is_bulk_loading;
```
Comment (hycdong) [inline src/common/bulk_load.thrift:209]: If `app_status` = BULK_LOAD::INVALID, it means this app is not executing bulk load, I think this param is redundant.

Comment (hycdong) [inline src/common/bulk_load.thrift:209]: Please update the `Possible errors` for `query_bulk_load_response` according to the function `on_query_bulk_load_status`. The wrong comments will raise worse misunderstanding.

File: src/meta/meta_bulk_load_service.cpp [hunk 6]
```diff
@@ -1402,17 +1466,16 @@ void bulk_load_service::on_query_bulk_load_status(query_bulk_load_rpc rpc)
...
+        response.err = (app == nullptr) ? ERR_APP_NOT_EXIST : ERR_APP_DROPPED;
         response.__set_hint_msg(hint_msg);
         return;
     }
 
     if (!app->is_bulk_loading) {
-        auto hint_msg = fmt::format("app({}) is not during bulk load", app_name);
+        auto hint_msg =
+            fmt::format("app({}) is not during bulk load, return last time result", app_name);
         derror_f("{}", hint_msg);
```
Comment (hycdong) [inline src/meta/meta_bulk_load_service.cpp:1477]: I think `dwarn_f` is better.

**Review by Shuo-Jia — DISMISSED** [review]

File: src/meta/meta_bulk_load_service.cpp (hunk 6 above)
Comment (GiantKing) [reply src/meta/meta_bulk_load_service.cpp:1477]: OK

File: src/common/bulk_load.thrift (hunk 5 above)
Comment (GiantKing) [reply src/common/bulk_load.thrift:209]: OK

**Review by hycdong — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**